x/evm/keeper/attest_reference_block.go

```diff
@@ -0,0 +1,66 @@
+package keeper
+
+import (
+	"context"
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/palomachain/paloma/x/consensus/keeper/consensus"
+	consensustypes "github.com/palomachain/paloma/x/consensus/types"
+	"github.com/palomachain/paloma/x/evm/types"
+)
+
+func (k Keeper) attestReferenceBlock(ctx context.Context, q consensus.Queuer, msg consensustypes.QueuedSignedMessageI) (retErr error) {
+	if len(msg.GetEvidence()) == 0 {
+		return nil
+	}
+
+	_, chainReferenceID := q.ChainInfo()
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	logger := k.Logger(ctx).WithFields(
+		"chain-id", chainReferenceID,
+		"component", "attest-reference-block",
+		"msg-id", msg.GetId(),
+		"msg-nonce", msg.Nonce())
+	logger.Debug("attest-reference-block")
+
+	cacheCtx, writeCache := sdkCtx.CacheContext()
+	defer func() {
+		if retErr == nil {
+			writeCache()
+		}
+	}()
+
+	result, err := k.consensusChecker.VerifyEvidence(cacheCtx, msg.GetEvidence())
+	if err != nil {
+		if errors.Is(err, ErrConsensusNotAchieved) {
+			logger.WithFields(
+				"total-shares", result.TotalShares,
+				"total-votes", result.TotalVotes,
+				"distribution", result.Distribution,
+			).WithError(err).Error("Consensus not achieved.")
+			return nil
+		}
+		return err
+	}
+
+	defer func() {
+		// given that there was enough evidence for a proof, regardless of the outcome,
+		// we should remove this from the queue as there isn't much that we can do about it.
+		if err := q.Remove(cacheCtx, msg.GetId()); err != nil {
+			k.Logger(sdkCtx).Error("error removing message, attestReferenceBlock", "msg-id", msg.GetId(), "msg-nonce", msg.Nonce())
```

Comment:
This as well. Can we maybe extract this kind of "house keeping" into something like a wrapper or orchestrator that governs all these attestations?

Reply:
I moved this and the previous into a wrapper and refactored the relevant functions.